### PR TITLE
Readd root/admin user detection

### PR DESCRIPTION
**Pull request by NoahvdAa**
This PR re-adds the root/admin user warning originally introduced in #2432. The original patch was reverted (ecfaff5) because of console spam issues on CentOS. In this PR, a different way of checking for for root/admin privileges is used, which shouldn't cause console spam on certain systems.

Tested on macOS 11.4, Ubuntu 20.04, Centos 8 and Windows Server 2022.

### CODE REVIEW

**Comment by molor** [comment]
In Windows Server, a single default user named Administrator is used on the every our host, and all user applications/services are run on his behalf (via RDP). We understand all the risks and agree with them, we take all the risks on ourselves. Where is the startup option that allows to bypass this (useless for us — me and my team) check and remove the warning?..

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 1]
```diff
@@ -0,0 +1,79 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: egg82 <eggys82@gmail.com>
```
Comment (Proximyst) [inline patches/server/0806-Add-root-admin-user-detection.patch]: Feel free to coauthor yourself as you came up with a new way of detecting it.

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
+index 22e9dd17f62103c5061435099ce96a3d70d54808..0c7247a3703f26ae3389daaa95ffbb30275c0125 100644
+--- a/src/main/java/org/bukkit/craftbukkit/Main.java
++++ b/src/main/java/org/bukkit/craftbukkit/Main.java
+@@ -261,6 +261,17 @@ public class Main {
+                     System.setProperty(TerminalConsoleAppender.JLINE_OVERRIDE_PROPERTY, "false"); // Paper
+                 }
+ 
++                // Paper start - detect running as root
++                if (io.papermc.paper.util.ServerEnvironment.userIsRootOrAdmin()) {
```
Comment (Proximyst) [inline patches/server/0806-Add-root-admin-user-detection.patch]: Let’s use a logger here

**Review by Proximyst — CHANGES_REQUESTED** [review]

**Review by Proximyst — CHANGES_REQUESTED** [review]
https://docs.oracle.com/en/java/javase/16/docs/api/jdk.security.auth/com/sun/security/auth/module/package-summary.html
There’s API to detect this properly, apparently. Could you please take a look at this instead?

**Comment by NoahvdAa** [comment]
Lynx and I came up with the solution used in 489bfa7, which uses Windows [Security identifiers](https://docs.microsoft.com/en-us/windows/security/identity-protection/access-control/security-identifiers). This seems to match command prompt running as administrator correctly. I'm not very familiar with how Windows works, can someone who knows Windows a little better help confirm that this is the best way to check for this?

**Comment by lynxplay** [comment]
> Lynx and I came up with the solution used in [489bfa7](https://github.com/PaperMC/Paper/commit/489bfa7dc0f912a4f15afaa0e0c677a25465b0f1), which uses Windows [Security identifiers](https://docs.microsoft.com/en-us/windows/security/identity-protection/access-control/security-identifiers). This seems to match command prompt running as administrator correctly. I'm not very familiar with how Windows works, can someone who knows Windows a little better help confirm that this is the best way to check for this?

Added to this, neither of us are running windows so if someone with access to a windows machine / windows-server machine could validate if this PR works correctly, that would be sweet :+1:

**Comment by TheFruxz** [comment]
MacOS should be checked too I think

**Comment by electronicboy** [comment]
it already does

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+             DedicatedServer.LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
+         }
+ 
++        // Paper start - detect running as root
++        if (io.papermc.paper.util.ServerEnvironment.userIsRootOrAdmin()) {
++            DedicatedServer.LOGGER.error("****************************");
++            DedicatedServer.LOGGER.error("YOU ARE RUNNING AS AN ADMINISTRATIVE OR ROOT USER. THIS IS NOT ADVISED.");
++            DedicatedServer.LOGGER.error("YOU ARE OPENING YOURSELF UP TO POTENTIAL RISKS WHEN DOING THIS.");
++            DedicatedServer.LOGGER.error("MALWARE, BAD PLUGINS, AND ATTACKERS WILL HAVE COMPLETE ACCESS AND CONTROL OF YOUR MACHINE.");
++            DedicatedServer.LOGGER.error("****************************");
```
Comment (e-im) [inline patches/server/0806-Add-root-admin-user-detection.patch]: > ATTACKERS WILL HAVE COMPLETE ACCESS AND CONTROL OF YOUR MACHINE

I really don't like the wording here, I can see a lot of people thinking that this means Minecraft "hackers" will be able to access their machine, playing further into the common misunderstanding that giving out your IP address will allow players access to your computer etc.

Comment (NoahvdAa) [reply patches/server/0806-Add-root-admin-user-detection.patch]: How does this sound?

> YOU ARE RUNNING THIS SERVER AS AN ADMINISTRATIVE OR ROOT USER. THIS IS NOT ADVISED.
> YOU ARE OPENING YOURSELF UP TO POTENTIAL RISKS WHEN DOING THIS.
> MALWARE AND MALICIOUS PLUGINS WILL HAVE COMPLETE ACCESS AND CONTROL OF YOUR MACHINE.

Comment (e-im) [reply patches/server/0806-Add-root-admin-user-detection.patch]: Yeah, just taking out the last bit about attackers makes it much better.

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
++public class ServerEnvironment {
++    private static final boolean runningAsRootOrAdmin;
++
++    static {
++        if (SystemUtils.IS_OS_WINDOWS) {
++            runningAsRootOrAdmin = Arrays.asList(new NTSystem().getGroupIDs()).contains("S-1-16-12288");
++        } else {
++            boolean isRunningAsRoot;
++            try (final var inputStream = Runtime.getRuntime().exec("id -u " + System.getProperty("user.name")).getInputStream()) {
++                isRunningAsRoot = new String(inputStream.readAllBytes()).trim().equals("0");
```
Comment (Proximyst) [inline patches/server/0806-Add-root-admin-user-detection.patch]: https://docs.oracle.com/en/java/javase/17/docs/api/jdk.security.auth/com/sun/security/auth/module/UnixSystem.html#getUid()

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
++
++import java.io.IOException;
++import java.util.Arrays;
++
++public class ServerEnvironment {
++    private static final boolean runningAsRootOrAdmin;
++
++    static {
++        if (SystemUtils.IS_OS_WINDOWS) {
++            runningAsRootOrAdmin = Arrays.asList(new NTSystem().getGroupIDs()).contains("S-1-16-12288");
```
Comment (Proximyst) [inline patches/server/0806-Add-root-admin-user-detection.patch]:
```suggestion
+            runningAsRootOrAdmin = Set.of(new NTSystem().getGroupIDs()).contains("S-1-16-12288");
```
Also move the constant out to a field to make it more readable.

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
++package io.papermc.paper.util;
++
++import com.sun.security.auth.module.NTSystem;
++import org.apache.commons.lang.SystemUtils;
++
++import java.io.IOException;
++import java.util.Arrays;
++
++public class ServerEnvironment {
++    private static final boolean runningAsRootOrAdmin;
```
Comment (Proximyst) [inline patches/server/0806-Add-root-admin-user-detection.patch]: `UPPER_SNAKE_CASE`

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -190,6 +190,16 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
+             DedicatedServer.LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
+         }
+ 
++        // Paper start - detect running as root
++        if (io.papermc.paper.util.ServerEnvironment.userIsRootOrAdmin()) {
++            DedicatedServer.LOGGER.error("****************************");
++            DedicatedServer.LOGGER.error("YOU ARE RUNNING THIS SERVER AS AN ADMINISTRATIVE OR ROOT USER. THIS IS NOT ADVISED.");
```
Comment (me4502) [inline patches/server/0806-Add-root-admin-user-detection.patch]: In order to minimise the amount of info about why this is a problem, could you please link to this post?

https://madelinemiller.dev/blog/root-minecraft-server/

I feel keeping the first line, then replacing the rest with "FOR MORE INFORMATION, SEE <link>" is probably the most concise wording. But if you have a better idea that also works

Comment (NoahvdAa) [reply patches/server/0806-Add-root-admin-user-detection.patch]: Replaced the third line with a link to your article. I feel like keeping the second line is important, to clearly signal that there is a risk.

File: patches/server/0806-Add-root-admin-user-detection.patch [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+index 1bf19965d12514dee34545235bfbadc0b74ddc8b..032f0fc0d8b7cd4142e008b184a071ac55ba7b32 100644
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -190,6 +190,16 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
+             DedicatedServer.LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
+         }
+ 
++        // Paper start - detect running as root
++        if (io.papermc.paper.util.ServerEnvironment.userIsRootOrAdmin()) {
++            DedicatedServer.LOGGER.error("****************************");
```
Comment (Proximyst) [inline patches/server/0806-Add-root-admin-user-detection.patch]: I'd argue error is the wrong level: this should be a `warn`.

**Review by Proximyst — APPROVED** [review]
Looks good

**Review by MiniDigger — APPROVED** [review]

**Pull request merged**